### ISSUE TITLE
force download an entire folder

### ISSUE THREAD

**Issue by zalmane** [Azure/azure-storage-fuse#1375]
### Which version of blobfuse was used?
2

### Which OS distribution and version are you using?
Ubuntu

### If relevant, please share your mount command.
sudo blobfuse2 mount /home/azureuser/blob --config-file=./blobfuse_config.yaml -o allow_other  --max-concurrency=500

### What was the issue encountered?
I am looking for a way to force download an entire folder, or to prefetch additional files. My application needs to read a folder of 2000 files. It has some level of parallelism in reading, but compared to AzCopy I am not able to get the same level of parallel downloads.
For comparison, I end up getting 5 minutes using Blobfuse2 vs 40 seconds using AzCopy. 
I can hack a threadpool to read 1 byte from each file, but looking to have BlobFuse do the heavy lifting by prefetching.
Is this possible? Any recommendations for additional config tweaks? 

### Please share logs if available.
Config.yaml:
```
components:
  - libfuse
  - file_cache
  - attr_cache
  - azstorage


libfuse:
  attribute-expiration-sec: 1200
  entry-expiration-sec: 1200
  negative-entry-expiration-sec: 240
  max-fuse-threads: 1000

file_cache:
  path: /home/azureuser/blobcache
  timeout-sec: 12000
  max-size-mb: 60960

attr_cache:
  timeout-sec: 7200

**Comment by vibhansa-msft** [comment]
As of now Blobfuse responds to file-system calls coming from kernel only. Unless application asks to open a file, blobfuse will not initiate download of the file. Also, if sufficient disk space is not available such prefetch downloads may fail.

We do see this ask coming from couple of customers where they want all of data to be prefetched into local disk for processing. We will try to prioritise this item in next couple of months. For now if application can fire a open call for all the files then all those files will get downloaded.

**Comment by zalmane** [comment]
Thank you for your response. Is it enough to initiate an open call or does
the application have to read at least one byte? What would be the easiest
way to initiate a mass open of all files in folder quickly that is
compliant with the way blobfuse is implemented?

**Comment by vibhansa-msft** [comment]
As you are using file-cache model, just opening the file will trigger the download of entire file. Based on the timeout you have configured files will remain in the temp storage. From application end you can just open the file and immediatly close it (without any read/write operations, just open in read-only mode may be)

**Comment by zalmane** [comment]
I am reading a parquet folder that has 2000 files. During the time the
parquet parses and loads the data of one fragment, blobfuse could read
ahead and download the next file. The difference would be several minutes.
I also tried two readers reading in parallel in a different order, but
seems I still only get about 10 concurrent downloads in the health monitor.
Is there some extra setting I can try?
Thank you!

**Comment by vibhansa-msft** [comment]
Blobfuse being a file-system driver does not know what is the next file application is interested in. To support such feature, like download entire folder or container, means we take a detour from traditional approach and just download stuff even if application does not request it. 

For paralllel downloads, you can just try out few threads just opening and closing the files in parallel. There is no hard limit on number of files you can open in parallel or download so shall be something coming from your application side and not from blobfuse2.

**Comment by zalmane** [comment]
The use case is a parquet reader that scans an entire parquet folder and filters just a small part of the data. It only tries to readahead 4 fragments so in practice, it spends most of the time waiting for the data to arrive.

**Comment by vibhansa-msft** [comment]
Does that mean in the folder you might have few hundred files, but you are interested only in a few files wich match certain criteria like size or last modified time or something? 
In the thread we were discussing about downloading the entire folder and if there is some level of filtering then downloading everything will be an overkill.

**Comment by zalmane** [comment]
Let me explain. Parquet "files" are typically a folder consisting of many parquet files, each consisting of a fragment.
Reading a full "Parquet" is possible with filters, which filters the data while reading it.
Let's say we have a "Parquet" consisting of 2000 fragments. We are reading in with a filter that only fetches certain row groups. The files are built so they can fetch only single row groups based on statistics stored in the header of the file. This allows file skipping and row group skipping.
So in many cases, the Parquet reader will only need to scan the header of the file, or a small part of the file.
Since it needs to wait for each file access to fully download, the whole process is slow.

On another note, I tried just opening and closing the file. This does achieve the result of faster parallel download. However, I tried just opening and closing the file, after I verified all files were in the cache, and still get ~12 seconds for 2000 open/close parallel calls (vs 0.1s when accessing the blobcache folder directly). Is that the expected behavior or is there a config flag to bypass some attribute lookups? All timeout values are using default, so the re-run is less than 120 seconds.

Thank you.

**Comment by vibhansa-msft** [comment]
When you open/close a file in quick succession it will just download the entire file in local cache and later when you try to read the file it will be served from local cache. As download of file is a network operation it will take some time (based on the network and backend locations etc). If application is generating open/close in parallel, then multiple files will be downloaded in parallel on local disk and will remain on disk till the timeout expires.

If you are not interested in full file contents, then you can also try out block-cache where files are downloaded only on need basis and not in entirety.

**Comment by zalmane** [comment]
The parallel open/close worked well. Compared to AZCOPY was a bit slower but within reason.
I couldn't use the block-cache since I am accessing ADLS.
I will close the Issue for now. Thank you for your prompt replies.

**Comment by vibhansa-msft** [comment]
ADLS and block-cache are not related in any way, block-cache is a blobfuse side component that helps you stream the files instead of downloading entire file in one shot.

**Comment by zalmane** [comment]
OK. I'll check again and open a separate issue with the error we got when
trying to access with block-cache enabled.
